test/javaTests/src/org/criu/java/tests/CheckpointRestore.java

```diff
@@ -0,0 +1,345 @@
+	if (null != logFiles) {
+		for (i = 0; i < logFiles.length; i++) {
+		deleteFile = new File(currentPath + logFiles[i]);
+		Files.delete(deleteFile.toPath());
+		}
+	}
+
+	boolean exists = Files.exists(Paths.get(currentPath + "stats-dump"));
+	if (exists) {
+		Files.delete(Paths.get(currentPath + "stats-dump"));
+	}
+
+	exists = Files.exists(Paths.get(currentPath + "stats-restore"));
+	if (exists) {
+		Files.delete(Paths.get(currentPath + "stats-restore"));
+	}
+	}
+
+	/**
+	 * Checkpoints the process, if process has been checkpointed correctly calls restore,
+	 * else makes a copy of dump.log, puts MappedBuffer to 'terminate' state and marks
+	 * test as failed
+	 *
+	 * @param pid Pid of process to be checkpointed
+	 * @throws IOException
+	 * @throws InterruptedException
+	 */
+	private void checkpoint(String pid) throws IOException, InterruptedException {
+	System.out.println("Checkpointing process " + pid);
+
+	ProcessBuilder builder = new ProcessBuilder("criu", "dump", "--tcp-established", "--shell-job", "-t", pid, "-vvv", "-D", logFolder, "-o", "dump.log");
+	Process process = builder.start();
+	BufferedReader stdError = new BufferedReader(new InputStreamReader(process.getErrorStream()));
+	int exitCode = process.waitFor();
+
+	if (0 != exitCode) {
+		/*
+		 * Print the error stream
+		 */
+		String line = stdError.readLine();
+		while (null != line) {
+		System.out.println(line);
+		line = stdError.readLine();
+		}
+
+		mappedByteBuffer.putChar(Helper.MAPPED_INDEX, Helper.STATE_TERMINATE);
+		/*
+		 * If checkpoint fails copy the dump.log
+		 */
+		Path fromPath = Paths.get(logFolder + "dump.log");
+		Path toPath = Paths.get(outputFolder + testName + "/" + "failedDump" + testName + ".log");
+		Files.copy(fromPath, toPath);
+		Assert.fail(testName + ": ERROR: Error during checkpoint: exitCode of checkpoint process was not zero.\nFor more details check " + outputFolder + testName + "/" + "failedDump" + testName + ".log");
+		return;
+	}
+
+	System.out.println("Checkpoint success");
+	process.destroy();
+
+	/*
+	 * The process has been checkpointed successfully, now restoring the process.
+	 */
+	try {
+		restore();
+	} catch (Exception e) {
+		mappedByteBuffer.putChar(Helper.MAPPED_INDEX, Helper.STATE_TERMINATE);
+		Assert.fail(testName + ": Exception occurred while restoring the test" + e, e);
+	}
+	}
+
+	/**
+	 * Restore the process, if process has been restored correctly put Mapped Buffer to
+	 * 'restored' state, else make a copy of restore.log and put MappedBuffer to 'terminate'
+	 * state and mark test as failed
+	 *
+	 * @throws IOException
+	 * @throws InterruptedException
+	 */
+	private void restore() throws IOException, InterruptedException {
+	System.out.println("Restoring process");
+
+	String currentPath = System.getProperty("user.dir");
+	ProcessBuilder builder = new ProcessBuilder(currentPath + "/../../scripts/criu-ns", "restore", "-d", "-vvv", "--tcp-established", "--shell-job", "-D", logFolder, "-o", "restore.log");
```

Comment:
Why do we need to use `criu-ns`? It is just a wrapper around criu that creates a new PID namespace.

Reply:
If I do not use criu-ns to restore the process in a new namespace, the restore fails because thread-id is not available( thread-id mismatch).
```
(00.027183) pie: 13541: Error (criu/pie/restorer.c:555): Thread pid mismatch 13541/13540
(00.027193) pie: 13541: Error (criu/pie/restorer.c:609): Restorer abnormal termination for 13532
(00.027316) Error (criu/cr-restore.c:2294): Restoring FAILED.
```

Reply:
Using `criu-ns` you create a new PID namespace and thus avoid PID collisions. Is the test runner part (a thread) of the process you are checkpointing/restoring?

If you are trying to checkpoint and restore yourself while still running this will not work. You either need to checkpoint/restore a separate process or include the restore in your startup code, but if there are still processes or threads running of the process you are trying to restore it will result in PID collisions. Restoring in a new PID namespace is a workaround, but I am not sure that is the actual thing you want to test.

Maybe some description in your test case what exactly you are trying to do here would be helpful.

Reply:
CheckpointRestore.java is our test runner, and it does not checkpoint itself. In CheckpointRestore.java, we run the test (FileRead.java) as a new process and it is this process we checkpoint and restore. The CheckpointRestore.java upon receiving a message from the test process checkpoints the test process, there is no "leave-running" option used when checkpointing the process. After the process has been checkpointed (and killed), CheckpointRestore.java restores the test process(currently in a new namespace using criu-ns).  So there is no thread that remains running.
What happens if after checkpointing, when the process has been killed, a thread-id that had been used by the test process gets used by some other process running on the machine. Won't the restore fail because the thread-id is not available.

Reply:
No, it does not always fail. It fails sometimes if we are not in a new namespace.

Reply:
> No, it does not always fail. It fails sometimes if we are not in a new namespace.

Do you know which process is blocking the restore? Some random process from the system or a process involved in the restore?

Reply:
We believe it is a random process, but need to confirm on that. 

Nidhi, can you retain the .img files and restore once again after confirming the thread ID is not in use.

Reply:
I restored the process again after ensuring the thread-id was not in use, and it restored successfully. It should be a random process, however I am unable to confirm that it was not a process involved in restore. Is there a way to do so? I am looking at the processes running on my system along with the thread-id of the threads used by them ( ps -elfT ) but I cannot find the thread-id which caused failure over there.

Reply:
@adrianreber - Are there any logging options that we can specify to log the process that was holding the thread-id when restore fails?